x/archivedb/key_value_reader.go

```diff
@@ -82,13 +28,49 @@ func (reader *dbHeightReader) Has(key []byte) (bool, error) {
 }
 
 // Get retrieves the given key if it's present in the key-value data store.
+//
+// This is a public API, so the dbKey is used to retrieve any value. It is
+// guaranteed to be a O(1), because how the dbKey is structured internally which
+// prevents a longer key to match a shorter key with the same prefix.
+//
+// If the result matches a dbMetaKey it will be consired as a ErrNotFound
+// because dbMetaKey is for internal usage and should be leaked outside of the
+// module
 func (reader *dbHeightReader) Get(key []byte) ([]byte, error) {
-	iterator, err := reader.getIteratorToValue(key)
+	iterator := reader.db.rawDB.NewIteratorWithStart(newDBKey(key, reader.height))
 	defer iterator.Release()
+
+	reader.heightLastFoundKey = 0
+
+	if !iterator.Next() {
+		err := iterator.Error()
+		if err != nil {
+			return nil, err
+		}
+
+		// There is no available key with the requested prefix
+		return nil, database.ErrNotFound
+	}
+
+	foundKey, height, err := parseDBKey(iterator.Key())
 	if err != nil {
 		return nil, err
 	}
-	return iterator.Value(), nil
+	if !bytes.Equal(foundKey, key) {
+		return nil, database.ErrNotFound
+	}
+	rawValue := iterator.Value()
+	rawValueLen := len(rawValue)
+	if rawValueLen == 0 {
+		// malformed, value should never ever be empty
```

Comment:
If this should never happen, let's return a different error

Reply:
This may very well happen, because we use the same keyspace to store metadata, such as the current height.,